### DISCUSSION TITLE
tuh_hid_set_report ASSERT fails after called many times

### DISCUSSION THREAD

**Discussion by morio** [hathach/tinyusb#1822] [Q&A]
I'm running an RP2040 as a USB host for a USB keyboard and mouse. I'm using `tuh_hid_set_report` to set the LEDs on the keyboard.  After hitting num lock or caps lock to change the LED status multiple times `tu_hid_set_report` starts failing an ASSERT in it's code. 
The tinyusb `tuh_task()` is being run on the second core, I also moved the code that use `tuh_hid_set_report` to the second core also in case of some RP2040 race condition. Here: https://github.com/rabbitholecomputing/QuokkADB-firmware/blob/612459f73dee975750d8d4709086de4c61809c46/src/firmware/lib/QuokkADB/src/quokkadb.cpp#L82 

I call `tuh_hid_set_report` here: https://github.com/rabbitholecomputing/QuokkADB-firmware/blob/612459f73dee975750d8d4709086de4c61809c46/src/firmware/lib/QuokkADB/src/keyboardrptparser.cpp#L153

This is the assert that fails: https://github.com/hathach/tinyusb/blob/7004914d8c6c7d88908776eb153661b6b987e513/src/class/hid/hid_host.c#L212

What causes the assert to fail is this line here:
[  TU_VERIFY(_ctrl_xfer.stage == CONTROL_STAGE_IDLE);](https://github.com/hathach/tinyusb/blob/7004914d8c6c7d88908776eb153661b6b987e513/src/host/usbh.c#L518)
At this point `_ctrl_xfer.stag` is equal to "CONTROL_STAGE_ACK"

Does this mean the keyboard is failing to complete the control endpoint handshake by not sending an ACK (don't know if the host or the device sends the ACK)?

Is it possible to call a function to force `_ctrl_xfer.stage` equal to `CONTROL_STAGE_IDLE`?

I've attached the log output from TinyUSB and the firmware I'm writing.

A normal key press to change the LEDs:
[working.log](https://github.com/hathach/tinyusb/files/10311223/working.log)

A normal key press to change LEDs right before it begins to fail:
[last_working_set_report.log](https://github.com/hathach/tinyusb/files/10311225/last_working_set_report.log)

A normal key press to change LEDs with the first instance of it failing
[first_failing_set_report.log](https://github.com/hathach/tinyusb/files/10311226/first_failing_set_report.log)

**Comment by hathach** [comment]
currently only 1 control transfer can be run at a time (no transfer queue support). I takes a usb frame or two to complete the transfer. If it returns false, meaning a transfer is on-going, you should try again later. Suggestion, block wait, or implement your own queue and execute transfer back-to-back with control xfer complete.
